src/context/puzzle.py

```diff
@@ -1,16 +1,73 @@
 from typing import List
 from src.models.puzzle import Puzzle
-from src.queries.puzzle import find_puzzles
+from src.queries.puzzle import get_puzzles, get_completed_puzzles
+from src.queries.player import get_player
 
 NUMBER_OF_FEEDERS = {"UTS": 4, "UNSW": 4, "USYD": 6}
 
 
-async def can_access_puzzle(puzzle: Puzzle, discord_id: int) -> bool:
-    # TODO: implementation
-    return True
+# the meat of the context. given a puzzle and a list of completed puzzle, determine if the puzzle is accessible yet
+# current logic:
+# all puzzles in a round are unlocked if the previous meta is completed (UTS is unlocked by default)
+# order is UTS -> USYD -> UNSW
+# solver must have solved all feeders to get the meta
+# all meta solved to get the metameta
+def can_access_puzzle_context(puzzle: Puzzle, completed_puzzles: List[Puzzle]) -> bool:
+    if puzzle.uni == "UTS" and puzzle.puzzle_id != "UTS-M":
+        return True
 
+    if (
+        puzzle.puzzle_id == "UTS-M"
+        and len([puzzle for puzzle in completed_puzzles if puzzle.uni == "UTS"])
+        >= NUMBER_OF_FEEDERS["UTS"]
+    ):
+        return True
 
-async def get_accessible_puzzles(discord_id: int) -> List[Puzzle]:
-    # TODO: implementation
-    puzzles = await find_puzzles()
-    return puzzles
+    if (
+        puzzle.uni == "USYD"
+        and puzzle.puzzle_id != "USYD-M"
+        and any([puzzle.puzzle_id == "UTS-M" for puzzle in completed_puzzles])
+    ):
+        return True
+
+    if (
+        puzzle.puzzle_id == "USYD-M"
+        and len([puzzles for puzzles in completed_puzzles if puzzle.uni == "USYD"])
+        >= NUMBER_OF_FEEDERS["USYD"]
+    ):
+        return True
+
+    if (
+        puzzle.uni == "UNSW"
+        and puzzle.puzzle_id != "UNSW-M"
+        and any([puzzle.puzzle_id == "USYD-M" for puzzle in completed_puzzles])
+    ):
+        return True
+
+    if (
+        puzzle.puzzle_id == "UNSW-M"
+        and len([puzzles for puzzles in completed_puzzles if puzzle.uni == "UNSW"])
+        >= NUMBER_OF_FEEDERS["UNSW"]
+    ):
+        return True
+
+    if puzzle.puzzle_id == "METAMETA" and any(
+        [puzzle.puzzle_id == "UNSW-M" for puzzle in completed_puzzles]
+    ):
+        return True
+
+    raise Exception("Invalid puzzle ID found!")
```

Comment:
This exception is never handled when the function is called.

Reply:
yup - code execution should not reach this point if puzzles have been entered correctly. I'll add more comments regarding this

Reply:
Okay sounds good.

Reply:
my bad - this was indeed a logic error. should be fixe